### ISSUE TITLE
catch_unwind doesn't catch panics from thread spawned inside the closure

### ISSUE THREAD

**Issue by mhoseinzadeh** [rust-lang/rust#72739]
Although the following example is poorly implemented, it would be nice if Rust disallowed `thread::spawn` from inside a `catch_unwind` because logically it might be impossible to catch the panic, unless the thread joins the main thread inside the closure.

```rust
fn thread_inside_catch_unwind() {
    std::panic::catch_unwind(|| {
        std::thread::spawn(|| {
            std::thread::sleep(std::time::Duration::from_secs(1));
            panic!()
        })
    }).unwrap();
}
```

**Comment by Mark-Simulacrum** [comment]
This is expected -- threads already catch unwinding inside them, see docs for the join method: https://doc.rust-lang.org/nightly/std/thread/struct.JoinHandle.html#method.join

**Comment by mhoseinzadeh** [comment]
Is it possible at all to disallow it? For example, a random function that takes a closure as its argument disallowing spawning threads?

**Comment by Mark-Simulacrum** [comment]
I don't think so, but I'm not sure I'm following you. Certainly you can e.g. have non-Send values in which case threads would not get spawned with that value. Generally Rust takes the approach that "threads are always there" though in terms of the guarantees it tries to offer.

If you have further questions, I would suggest users.rust-lang.org -- there's more people there who can help with questions like this and we generally try to keep the issue tracker just to bugs rather than questions.
